File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, 1, 0, -1};
int n, m;
bool out(int x, int y) {
    return x < 0 || x >= n || y < 0 || y >= m;
}

int solution(vector<string> board) {
    pair<int, int> R, G;
    n = board.size();
    m = board[0].size();
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (board[i][j] == 'R') R = {i, j};
            if (board[i][j] == 'G') G = {i, j};
        }
    }
    
    vector<vector<int>> dist(n, vector<int>(m, -1));
    dist[R.first][R.second] = 0;
    queue<pair<int, int>> q;
    q.push(R);
    
    while(!q.empty()) {
        auto cur = q.front(); q.pop();
        int x = cur.first, y = cur.second;
        
        if (cur == G) return dist[x][y];
        
        for (int i = 0; i < 4; i++) {
            int nx = x;
            int ny = y;
            
            while(true) {
                int tx = nx + dx[i];
                int ty = ny + dy[i];
                
                if (out(tx, ty) || board[tx][ty] == 'D') break;
                nx = tx;
                ny = ty;
            }
            
            if (nx == x && ny == y) continue;
            if (dist[nx][ny] != -1) continue;
            
            dist[nx][ny] = dist[x][y] + 1;
            q.push({nx, ny});
        }
    }
    
    
    return -1;
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;
class Solution {
    public int solution(String[] board) {
        int[] r = new int[2];
        int[] g = new int[2];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length(); j++) {
                if (board[i].charAt(j) == 'R') {
                    r = new int[]{i, j};
                }
                if (board[i].charAt(j) == 'G') {
                    g = new int[]{i, j};
                }
            }
        }
        
        return bfs(r, g, board);
    }
    
    int[] dx = {1, -1, 0, 0};
    int[] dy = {0, 0, 1, -1};
    
    private int bfs(int[] r, int[] g, String[] board) {
        int n = board.length;
        int m = board[0].length();
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];
        
        q.offer(new int[]{r[0], r[1], 0});
        visited[r[0]][r[1]] = true;
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            
            if (cur[0] == g[0] && cur[1] == g[1]) return cur[2];
            
            for (int d = 0; d < 4; d++) {
                int nx = cur[0];
                int ny = cur[1];
                                
                while (true) {
                    int nextX = nx + dx[d];
                    int nextY = ny + dy[d];
                    
                    if (nextX < 0 || nextY < 0 || nextX >= n || nextY >= m) break;
                    if (board[nextX].charAt(nextY) == 'D') break;

                    nx = nextX;
                    ny = nextY;
                }
                
                if (visited[nx][ny]) continue;

                visited[nx][ny] = true;
                q.offer(new int[]{nx, ny, cur[2] + 1});
            }
        }

        return -1;
    }
}